raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java

```diff
@@ -0,0 +1,271 @@
+        int localId = 0;
+        int leader = 1;
+        Uuid localDirectoryId = Uuid.randomUuid();
+        Uuid leaderDirectoryId = Uuid.randomUuid();
+        Set<Integer> voterIds = new HashSet<>(Arrays.asList(localId, leader));
+        Set<ReplicaKey> voters = new HashSet<>(Arrays.asList(
+            ReplicaKey.of(localId, localDirectoryId),
+            ReplicaKey.of(leader, leaderDirectoryId)));
+        VoterSet voterSet = VoterSetTest.voterSet(voters.stream());
+        short kRaftVersion = 1;
+        int epoch = 0;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(OptionalInt.of(localId), voterIds, localDirectoryId, kRaftVersion)
+            .withKip853Rpc(withKip853Rpc)
+            .withElectedLeader(epoch, leader)
+            .withBootstrapSnapshot(Optional.of(voterSet), kRaftVersion)
+            .build();
+
+        // check that follower will send fetch request to leader
+        context.pollUntilRequest();
+        RaftRequest.Outbound fetchRequest = context.assertSentFetchRequest();
+        context.assertFetchRequestData(fetchRequest, epoch, 0L, 0);
+
+        // check if leader response were to contain bootstrap snapshot id, follower would not send fetch snapshot request
+        context.deliverResponse(
+            fetchRequest.correlationId(),
+            fetchRequest.destination(),
+            context.snapshotFetchResponse(epoch, leader, BOOTSTRAP_SNAPSHOT_ID, 0)
+        );
+        context.pollUntilRequest();
+        fetchRequest = context.assertSentFetchRequest();
+        context.assertFetchRequestData(fetchRequest, epoch, 0L, 0);
+    }
+
+    // this test is maybe out of scope - we are checking that follower is able to apply votersRecord essentially
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testFollowerReadsKRaftBootstrapRecords(boolean withKip853Rpc) throws Exception {
+        int localId = 0;
+        int leader = 1;
+        int follower2 = 2;
+        Uuid localDirectoryId = Uuid.randomUuid();
+        Uuid leaderDirectoryId = Uuid.randomUuid();
+        Uuid followerDirectoryId2 = Uuid.randomUuid();
+        Set<Integer> voterIds = new HashSet<>(Arrays.asList(localId, leader));
+        Set<ReplicaKey> voters = new HashSet<>(Arrays.asList(
+            ReplicaKey.of(localId, localDirectoryId),
+            ReplicaKey.of(leader, leaderDirectoryId)));
+        VoterSet voterSet = VoterSetTest.voterSet(voters.stream());
+        short kRaftVersion = 1;
+        int epoch = 5;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(OptionalInt.of(localId), voterIds, localDirectoryId, kRaftVersion)
+            .withKip853Rpc(withKip853Rpc)
+            .withElectedLeader(epoch, leader)
+            .withBootstrapSnapshot(Optional.of(voterSet), kRaftVersion)
+            .appendToLog(epoch, Arrays.asList("a", "b", "c"))
```

Comment:
Technically this is not a valid state. The whole point of your PR is that the start of the log should contain the control records (KRaftVersionRecord and VotersRecord) from bootstrap checkpoint before any other record.

Reply:
nevermind, I see why you're saying this is invalid - the follower replicates from the leader which should not have other records before these control records